extensions/amp-ad/0.1/amp-ad-custom.js

```diff
@@ -0,0 +1,145 @@
+/**
+ * Copyright 2016 David Karlin, Bachtrack Ltd. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {isLayoutSizeDefined} from '../../../src/layout';
+import {user} from '../../../src/log';
+import {templatesFor} from '../../../src/template';
+import {xhrFor} from '../../../src/xhr';
+import {addParamToUrl} from '../../../src/url';
+import {ancestorElementsByTag} from '../../../src/dom';
+
+/** @const {!string} Tag name for custom ad implementation. */
+export const TAG_AD_CUSTOM = 'amp-ad-custom';
+
+/** @var {Object} A map of promises for each value of data-url. The promise
+ *  will fetch data for the URL for the ad server, and return it as a map of
+ *  objects, keyed by slot; each object contains the variables to be
+ *   substituted into the mustache template. */
+const ampCustomadXhrPromises = {};
+
+/** @var {Object} a map of full urls (i.e. including the ampslots parameter)
+ * for each value of data-url */
+let ampCustomadFullUrls = null;
+
+export class AmpAdCustom extends AMP.BaseElement {
+
+  /** @param {!AmpElement} element */
+  constructor(element) {
+    super(element);
+    /** @private {string} The base URL of the ad server for this ad */
+    this.url_ = element.getAttribute('data-url');
+
+    /** @private {string} A string identifying this ad slot: the server's
+     *  responses will be keyed by slot */
+    this.slot_ = element.getAttribute('data-slot');
+
+    /** @private {boolean} True to suppress template checks and rendering:
+     * use this during testing, since our test harness does not yet support
+     * templates.
+     */
+    this.testMode_ = element.getAttribute('data-test');
+  }
+
+  /** @override */
+  getPriority() {
+    // Loads ads after other content.
+    return 2;
+  }
+
+  /** @override **/
+  isLayoutSupported(layout) {
+    /** @TODO Add proper support for more layouts, and figure out which ones
+     *  we're permitting */
+    return isLayoutSizeDefined(layout);
+  }
+
+  buildCallback() {
+    // Ensure that there are templates in this ad
+    const templates = this.element.querySelectorAll('template');
+    if (!this.testMode_) {
+      user().assert(templates.length > 0, 'Missing template in custom ad');
+    }
+    // And ensure that the slot value is legal
+    user().assert(this.slot_ === null || this.slot_.match(/^[0-9a-z]+$/),
+        'custom ad slot should be alphanumeric: ' + this.slot_);
+  }
```

Comment:
I forgot one thing, you will need a `uiHandler` here, like in `amp-ad-3p-impl.js`:

```
this.uiHandler = new AmpAdUIHandler(this);
this.uiHandler.init();
```